internal/scheduler/scheduler_test.go

```diff
@@ -0,0 +1,449 @@
+					Status: argov1alpha1.ApplicationStatus{
+						Sync: argov1alpha1.SyncStatus{
+							Status: argov1alpha1.SyncStatusCodeOutOfSync,
+						},
+					},
+				},
+			},
+			stage: deploymentskyscannernetv1alpha1.ProgressiveRolloutStage{
+				Name:        "",
+				MaxParallel: intstr.IntOrString{IntVal: 1},
+				MaxTargets:  intstr.IntOrString{IntVal: 1},
+				Targets:     deploymentskyscannernetv1alpha1.ProgressiveRolloutTargets{},
+			},
+			expected: []string{
+				"app-one",
+			},
+		},
+		// 5 Applications:
+		//  - OutOfSync 5
+		//  - syncedInCurrentStage 0
+		//  - Progressing 0
+		// Stage:
+		//  - maxTargets: 3
+		//  - maxParallel: 3
+		// The scheduler should return 3 applications
+		{
+			apps: []argov1alpha1.Application{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "app-one",
+						Namespace: namespace,
+					},
+					Status: argov1alpha1.ApplicationStatus{
+						Sync: argov1alpha1.SyncStatus{
+							Status: argov1alpha1.SyncStatusCodeOutOfSync,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "app-two",
+						Namespace: namespace,
+					},
+					Status: argov1alpha1.ApplicationStatus{
+						Sync: argov1alpha1.SyncStatus{
+							Status: argov1alpha1.SyncStatusCodeOutOfSync,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "app-three",
+						Namespace: namespace,
+					},
+					Status: argov1alpha1.ApplicationStatus{
+						Sync: argov1alpha1.SyncStatus{
+							Status: argov1alpha1.SyncStatusCodeOutOfSync,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "app-four",
+						Namespace: namespace,
+					},
+					Status: argov1alpha1.ApplicationStatus{
+						Sync: argov1alpha1.SyncStatus{
+							Status: argov1alpha1.SyncStatusCodeOutOfSync,
+						},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "app-five",
+						Namespace: namespace,
+					},
+					Status: argov1alpha1.ApplicationStatus{
+						Sync: argov1alpha1.SyncStatus{
+							Status: argov1alpha1.SyncStatusCodeOutOfSync,
+						},
+					},
+				},
+			},
+			stage: deploymentskyscannernetv1alpha1.ProgressiveRolloutStage{
+				Name:        "test-5",
+				MaxParallel: intstr.IntOrString{IntVal: 3},
+				MaxTargets:  intstr.IntOrString{IntVal: 3},
+				Targets:     deploymentskyscannernetv1alpha1.ProgressiveRolloutTargets{},
+			},
+			expected: []string{
+				"app-five", "app-four", "app-one",
+			},
+		},
+		// 0 Applications
+		// Stage:
+		//  - maxTargets: 3
+		//  - maxParallel: 3
+		// The scheduler should return 0 applications
+		{
+			apps: nil,
+			stage: deploymentskyscannernetv1alpha1.ProgressiveRolloutStage{
+				Name:        "test-5",
+				MaxParallel: intstr.IntOrString{IntVal: 3},
+				MaxTargets:  intstr.IntOrString{IntVal: 3},
+				Targets:     deploymentskyscannernetv1alpha1.ProgressiveRolloutTargets{},
+			},
+			expected: nil,
+		},
+	}
+
+	for _, testCase := range testCases {
```

Comment:
The reason why I went with that structure it's because I saw that in multiple places. For example:
- https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/pkg/backend/endpoint_utils_test.go
- https://github.com/argoproj-labs/applicationset/blob/master/pkg/generators/cluster_test.go
- https://github.com/fluxcd/helm-controller/blob/main/controllers/helmrelease_controller_chart_test.go
- https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/alertmanager/amcfg_test.go

and I'm pretty sure there are others.
I'm happy to have a chat about if we want to keep this style or change it. Do you have an example repository or a test?

> Having a for loop means that the later assertions don't run if the first fails, for instance.

True, but you still know you have at least a test to fix :)

> Individual test names can also be helpful for readability.

I thought about adding a `name` field to the test struct, but then I considered how long the description is and I thought about adding a comment instead. Happy to add a name if we think it helps with the readability.

Reply:
Added the `name` field to the Scheduler tests. The output looks like
```
=== RUN   TestScheduler
Test:  Applications: outOfSync 3, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 2, maxParallel 2 | Expected: scheduled 2
Test:  Applications: outOfSync 3, syncedInCurrentStage 1, progressing 1, | Stage: maxTargets 5, maxParallel 2 | Expected: scheduled 1
Test:  Applications: outOfSync 5, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 3, maxParallel 2 | Expected: scheduled 2
Test:  Applications: outOfSync 5, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 50%, maxParallel 100% | Expected: scheduled 2
Test:  Applications: outOfSync 2, syncedInCurrentStage 3, progressing 0, | Stage: maxTargets 3, maxParallel 1 | Expected: scheduled 0
Test:  Applications: outOfSync 1, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 1, maxParallel 1 | Expected: scheduled 1
Test:  Applications: outOfSync 5, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 3, maxParallel 3 | Expected: scheduled 3
Test:  Applications: outOfSync 2, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 10%, maxParallel 10% | Expected: scheduled 1
Test:  Applications: outOfSync 2, syncedInCurrentStage 0, progressing 0, syncedInPreviousStage 1 | Stage: maxTargets 1, maxParallel 1 | Expected: scheduled 1
Test:  Applications: outOfSync 0, syncedInCurrentStage 0, progressing 0, | Stage: maxTargets 3, maxParallel 3 | Expected: scheduled 0
--- PASS: TestScheduler (0.00s)
```
What do you think?

Reply:
Using `t.Run` seems a little better
```
=== RUN   TestScheduler
=== RUN   TestScheduler/Applications:_outOfSync_3,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_2,_maxParallel_2_|_Expected:_scheduled_2
=== RUN   TestScheduler/Applications:_outOfSync_3,_syncedInCurrentStage_1,_progressing_1,_|_Stage:_maxTargets_5,_maxParallel_2_|_Expected:_scheduled_1
=== RUN   TestScheduler/Applications:_outOfSync_5,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_2_|_Expected:_scheduled_2
=== RUN   TestScheduler/Applications:_outOfSync_5,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_50%,_maxParallel_100%_|_Expected:_scheduled_2
=== RUN   TestScheduler/Applications:_outOfSync_2,_syncedInCurrentStage_3,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_1_|_Expected:_scheduled_0
=== RUN   TestScheduler/Applications:_outOfSync_1,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_1,_maxParallel_1_|_Expected:_scheduled_1
=== RUN   TestScheduler/Applications:_outOfSync_5,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_3_|_Expected:_scheduled_3
=== RUN   TestScheduler/Applications:_outOfSync_2,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_10%,_maxParallel_10%_|_Expected:_scheduled_1
=== RUN   TestScheduler/Applications:_outOfSync_2,_syncedInCurrentStage_0,_progressing_0,_syncedInPreviousStage_1_|_Stage:_maxTargets_1,_maxParallel_1_|_Expected:_scheduled_1
=== RUN   TestScheduler/Applications:_outOfSync_0,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_3_|_Expected:_scheduled_0
--- PASS: TestScheduler (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_3,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_2,_maxParallel_2_|_Expected:_scheduled_2 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_3,_syncedInCurrentStage_1,_progressing_1,_|_Stage:_maxTargets_5,_maxParallel_2_|_Expected:_scheduled_1 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_5,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_2_|_Expected:_scheduled_2 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_5,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_50%,_maxParallel_100%_|_Expected:_scheduled_2 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_2,_syncedInCurrentStage_3,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_1_|_Expected:_scheduled_0 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_1,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_1,_maxParallel_1_|_Expected:_scheduled_1 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_5,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_3_|_Expected:_scheduled_3 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_2,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_10%,_maxParallel_10%_|_Expected:_scheduled_1 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_2,_syncedInCurrentStage_0,_progressing_0,_syncedInPreviousStage_1_|_Stage:_maxTargets_1,_maxParallel_1_|_Expected:_scheduled_1 (0.00s)
    --- PASS: TestScheduler/Applications:_outOfSync_0,_syncedInCurrentStage_0,_progressing_0,_|_Stage:_maxTargets_3,_maxParallel_3_|_Expected:_scheduled_0 (0.00s)
PASS
```
because the result is automatically broken down by the individual test case.

Reply:
The formatting is not great because it's replacing a space with `_`, so we might need to change the test description :)

Reply:
Right - this seems to be recommended by Golang in any case (https://github.com/golang/go/wiki/TableDrivenTests).

There may be value in something like https://onsi.github.io/ginkgo/#table-driven-tests, which seems to take the input/expected and generate different independent test cases; but with this being common in other projects and the output being descriptive as to 'which' test failed, I'm not sure if it's worth it to change.